File: bindings/kotlin/src/main/kotlin/strling/Compiler.kt
package strling

import com.strling.jvm.NativeClient

/** Immutable options for constructing canonical source CompileRequests. */
data class SourceCompileOptions(
    val sourceId: String = "src:kotlin.adapter",
    val frontendId: String = "semantic_strling",
    val frontendVersion: String? = null,
    val mediaType: String = if (frontendId == "legacy_regex") "text/x-regex" else "text/strling",
    val specificationVersion: String = "1.0-draft.1",
    val requestedOutputs: List<String>? = null,
    val compilerOptions: Map<String, Any?> = defaultCompilerOptions(),
    val targetProfileReference: Map<String, Any?>? = null,
    val targetProfile: Map<String, Any?>? = null,
) {
    companion object {
        @JvmStatic
        fun defaultCompilerOptions(): Map<String, Any?> = mapOf(
            "partial_semantics" to "forbid",
            "diagnostic_policy" to mapOf("minimum_severity" to "hint"),
        )
    }
}

/** Canonical compile-request conveniences over a supplied native client. */
class Compiler(private val client: NativeClient) {
    fun compile(request: Map<String, *>, targetProfile: Map<String, *>? = null): Any? =
        client.compile(request, targetProfile)

    fun check(request: Map<String, *>, targetProfile: Map<String, *>? = null): Any? =
        compile(request, targetProfile)

    /** Compatibility name returning canonical compile data, never a local AST. */
    fun parse(source: String, options: SourceCompileOptions = SourceCompileOptions()): Any? =
        client.compile(
            sourceCompileRequest(source, options, listOf("semantic")),
            options.targetProfile,
        )

    /** Compatibility name requesting a canonical TargetArtifact. */
    fun parseToArtifact(source: String, options: SourceCompileOptions): Any? {
        require(options.targetProfile != null && options.targetProfileReference != null) {
            "parseToArtifact requires an exact target profile and reference"
        }
        return client.compile(
            sourceCompileRequest(
                source,
                options,
                listOf("semantic", "portability", "target_artifact"),
            ),
            options.targetProfile,
        )
    }

    fun describe(): Any? = client.describe()

    fun inspectTargetProfile(targetProfile: Map<String, *>): Any? =
        client.inspectTargetProfile(targetProfile)

    companion object {
        @JvmStatic
        fun sourceCompileRequest(
            source: String,
            options: SourceCompileOptions = SourceCompileOptions(),
            fallbackOutputs: List<String> = listOf("semantic", "analysis"),
        ): Map<String, Any?> = mapOf(
            "contract_version" to "1.0.0",
            "specification_version" to options.specificationVersion,
            "input" to mapOf(
                "kind" to "source",
                "document" to mapOf(
                    "contract_version" to "1.0.0",
                    "source_id" to options.sourceId,
                    "specification_version" to options.specificationVersion,
                    "frontend" to mapOf(
                        "id" to options.frontendId,
                        "dialect_version" to (options.frontendVersion ?: options.specificationVersion),
                    ),
                    "content" to mapOf(
                        "kind" to "inline",
                        "encoding" to "utf-8",
                        "media_type" to options.mediaType,
                        "text" to source,
                    ),
                    "provenance" to mapOf("kind" to "authored"),
                ),
            ),
            "requested_outputs" to (options.requestedOutputs ?: fallbackOutputs),
            "compiler_options" to options.compilerOptions,
        ).let { request ->
            options.targetProfileReference?.let {
                request + ("target_profile" to it)
            } ?: request
        }
    }
}

File: bindings/python/src/STRling/compiler.py
"""Canonical compile-request conveniences over a supplied native client."""

from __future__ import annotations

from typing import Any, Mapping, Optional, Sequence

from STRling.interop import NativeClient


class Compiler:
    def __init__(self, client: NativeClient) -> None:
        self._client = client

    def compile(
        self,
        request: Mapping[str, Any],
        target_profile: Optional[Mapping[str, Any]] = None,
    ) -> Any:
        return self._client.compile(request, target_profile)

    def check(
        self,
        request: Mapping[str, Any],
        target_profile: Optional[Mapping[str, Any]] = None,
    ) -> Any:
        return self.compile(request, target_profile)


def parse(
    client: NativeClient,
    source: str,
    options: Optional[Mapping[str, Any]] = None,
) -> Any:
    prepared = dict(options or {})
    return client.compile(
        source_compile_request(source, prepared, ("semantic",)),
        prepared.get("target_profile"),
    )


def parse_to_artifact(
    client: NativeClient,
    source: str,
    options: Mapping[str, Any],
) -> Any:
    prepared = dict(options)
    target_profile = prepared.get("target_profile")
    target_reference = prepared.get("target_profile_reference")
    if not isinstance(target_profile, Mapping) or not isinstance(
        target_reference, Mapping
    ):
        raise TypeError(
            "parse_to_artifact requires an exact target profile and reference"
        )
    return client.compile(
        source_compile_request(
            source,
            prepared,
            ("semantic", "portability", "target_artifact"),
        ),
        target_profile,
    )


def source_compile_request(
    source: str,
    options: Optional[Mapping[str, Any]] = None,
    fallback_outputs: Sequence[str] = ("semantic", "analysis"),
) -> dict[str, Any]:
    prepared = dict(options or {})
    specification_version = str(prepared.get("specification_version", "1.0-draft.1"))
    frontend_id = str(prepared.get("frontend_id", "semantic_strling"))
    request = {
        "contract_version": "1.0.0",
        "specification_version": specification_version,
        "input": {
            "kind": "source",
            "document": {
                "contract_version": "1.0.0",
                "source_id": str(prepared.get("source_id", "src:python.adapter")),
                "specification_version": specification_version,
                "frontend": {
                    "id": frontend_id,
                    "dialect_version": str(
                        prepared.get("frontend_version", specification_version)
                    ),
                },
                "content": {
                    "kind": "inline",
                    "encoding": "utf-8",
                    "media_type": str(
                        prepared.get(
                            "media_type",
                            "text/x-regex"
                            if frontend_id == "legacy_regex"
                            else "text/strling",
                        )
                    ),
                    "text": source,
                },
                "provenance": {"kind": "authored"},
            },
        },
        "requested_outputs": list(prepared.get("requested_outputs", fallback_outputs)),
        "compiler_options": dict(
            prepared.get(
                "compiler_options",
                {
                    "partial_semantics": "forbid",
                    "diagnostic_policy": {"minimum_severity": "hint"},
                },
            )
        ),
    }
    target_reference = prepared.get("target_profile_reference")
    if isinstance(target_reference, Mapping):
        request["target_profile"] = dict(target_reference)
    return request


__all__ = [
    "Compiler",
    "parse",
    "parse_to_artifact",
    "source_compile_request",
]
